pkg/util/git.go

```diff
@@ -0,0 +1,203 @@
+			}
+		}
+	}
+
+	log.Printf("Could not find branch %s", branch)
+
+	return false, nil
+}
+
+// TODO: Need to handle https and ssh auth sanely
+func SyncRepo(repoURL, destination string) error {
+	log.Printf("Syncing %s to %s:", repoURL, destination)
+
+	if _, err := os.Stat(destination); !os.IsNotExist(err) {
+		chdirErr := os.Chdir(destination)
+		if chdirErr != nil {
+			return chdirErr
+		}
+
+		repo, repoErr := git.PlainOpen(destination)
+		if repoErr != nil {
+			return repoErr
+		}
+
+		w, err := repo.Worktree()
+		//CheckIfError(err)
+		if err != nil {
+			return err
+		}
+
+		// ... checking out to commit
+		//Info("git checkout %s", commit)
+		err = w.Checkout(&git.CheckoutOptions{
+			Branch: plumbing.NewBranchReferenceName("master"),
+		})
+		//CheckIfError(err)
+		if err != nil {
+			return err
+		}
+
+		pullOpts := &git.PullOptions{RemoteName: "origin"}
+		err = w.Pull(pullOpts)
+		if err != nil && err != git.NoErrAlreadyUpToDate {
+			return err
+		}
+
+		return nil
+	}
+
+	cloneOpts := &git.CloneOptions{
+		URL: repoURL,
+	}
+	_, repoErr := git.PlainClone(destination, false, cloneOpts)
+	if repoErr != nil {
+		return repoErr
+	}
+
+	return nil
+}
+
+// RevParse parses a git revision and returns a SHA1 on success, otherwise an
+// error.
+func RevParse(rev, workDir string) (string, error) {
+	repo, err := git.PlainOpen(workDir)
+	if err != nil {
+		return "", err
+	}
+
+	ref, err := repo.ResolveRevision(plumbing.Revision(rev))
+	if err != nil {
+		return "", err
+	}
+
+	return ref.String(), nil
+}
+
+// RevParseShort parses a git revision and returns a SHA1 on success, otherwise an
+// error.
+func RevParseShort(rev, workDir string) (string, error) {
+	fullRev, err := RevParse(rev, workDir)
+	if err != nil {
+		return "", err
+	}
+
+	shortRev := fullRev[:10]
+
+	return shortRev, nil
+}
+
+func GetMergeBase(masterRefShort, releaseRefShort string, repo *git.Repository) (string, error) {
+	masterRef := fmt.Sprintf("origin/%s", masterRefShort)
+	releaseRef := fmt.Sprintf("origin/%s", releaseRefShort)
+
+	log.Printf("masterRef: %s, releaseRef: %s", masterRef, releaseRef)
+
+	commitRevs := []string{masterRef, releaseRef}
+	var res []*object.Commit
+
+	var hashes []*plumbing.Hash
+	for _, rev := range commitRevs {
+		hash, err := repo.ResolveRevision(plumbing.Revision(rev))
+		//checkIfError(err, exitCodeCouldNotParseRevision, "could not parse revision '%s'", rev)
+		if err != nil {
+			return "", err
+		}
+		hashes = append(hashes, hash)
+	}
+
+	var commits []*object.Commit
+	for _, hash := range hashes {
+		commit, err := repo.CommitObject(*hash)
+		//checkIfError(err, exitCodeUnexpected, "could not find commit '%s'", hash.String())
+		if err != nil {
+			return "", err
+		}
+		commits = append(commits, commit)
+	}
+
+	res, err := commits[0].MergeBase(commits[1])
+	//checkIfError(err, exitCodeUnexpected, "could not traverse the repository history")
+	if err != nil {
+		return "", err
+	}
+
+	if len(res) == 0 {
+		os.Exit(1)
```

Comment:
I suggest exit'ing in the main path and return an error here instead.

Reply:
Fixed!